src/elevator/Elevator.java

```diff
@@ -4,91 +4,122 @@
-	public void run() {
-		try {
-			while (true) {
-				byte data[] = new byte[100];
-				DatagramPacket receivePacket = receive(data);
-
-				DatagramPacket sendPacket = new DatagramPacket(data,
-						data.length, receivePacket.getAddress(),
-						receivePacket.getPort());
-
-				send(sendPacket);
-			}
-		} catch (Exception e) {
-			System.out.println("Elevator quiting.");
-			e.printStackTrace();
-			close();
-		}
-	}
-
-	public void close() {
-		sendSocket.close();
-		receiveSocket.close();
-	}
-
-	public static void main(String args[]) {
-		System.out.println("Elevator: Starting on port 4000");
-		Elevator s = new Elevator();
-		s.run();
-	}
+    // State is the possible states that the elevator can be in.
+    enum State {
+        MOVING_UP,
+        MOVING_DOWN,
+        STOPPED_DOORS_CLOSED,
+        STOPPED_DOORS_OPENED
+    }
+
+    private State state;
+
+    // Simulates the physical attached with the elevator.
+    private Motor motor;
+    private Door door;
+
+    // Communication sockets
+    private DatagramSocket sendSocket, receiveSocket;
+
+    public Elevator() {
+        state = State.STOPPED_DOORS_CLOSED;
+        motor = new Motor();
+        door = new Door();
+
+        try {
+            sendSocket = new DatagramSocket();
+            receiveSocket = new DatagramSocket(4000);
+        } catch (SocketException se) {
+            se.printStackTrace();
+            System.exit(1);
+        }
+    }
+
+    /*
+     * run starts the elevator subsystem and awaits for messages.
+     */
+    public void run() {
+        try {
+            while (true) {
+                byte data[] = new byte[100];
+                DatagramPacket receivePacket = Message.receive(receiveSocket, data);
+                Message m = Message.deserialize(receivePacket.getData());
+                handleMessage(m);
+            }
+        } catch (Exception e) {
+            System.out.println("Elevator quiting.");
+            e.printStackTrace();
+            close();
+        }
+    }
+
+    /*
+     * close closes the communication sockets of the elevator subsystem.
+     */
+    public void close() {
+        sendSocket.close();
+        receiveSocket.close();
+    }
+
+    /*
+     * handleMessage runs the corresponding action for the message type.
+     */
+    private void handleMessage(Message m) {
+        // State machine switch
+        if (m instanceof ElevatorMessage) {
+            // scheduler tells the elevator to move, stop, open or close.
+            handleElevatorMessage((ElevatorMessage) m);
+            // } else if (m instanceof FloorMessage) {
+            //     reroute the message to the schduler
+            }
+    }
+
+    /*
+     * handleElevatorMessage handles commands for the elevator's physical components.
+     */
+    private void handleElevatorMessage(ElevatorMessage m) {
+        switch (m.getMessageType()) {
+            case STOP:
+                assert(state == State.MOVING_UP || state == State.MOVING_DOWN);
```

Comment:
These assertions are going to help up debug ensure that our elevator state machine is following the path we expect. I'm not sure how we can handle out of order messages.

Reply:
Assume a ideal network

Reply:
we're already using udp so I guess this is the correct way forward